runtime/src/main/java/org/corfudb/runtime/view/ManagementView.java

```diff
@@ -320,125 +238,303 @@ private NodeStatus getLogUnitNodeStatusInLayout(Layout layout, String server) {
+                    return getUnavailableClusterStatusReport(layoutServers, ClusterStatusReliability.WEAK_NO_QUORUM, layoutMap.keySet(), layout);
+                }
+
+                log.debug("getClusterStatus: quorum layout {}", layout);
             }
+
+            // Get Cluster Status from Layout
+            ClusterStatus clusterStatus = getClusterHealth(layout, layout.getAllActiveServers());
+
+            // Get Node Status from Layout
+            Map<String, NodeStatus> nodeStatusMap = getNodeStatusMap(layout);
+
+            // To complete cluster status with connectivity information of this
+            // client to every node in the cluster, ping all servers
+            // TODO: we could eventually skip pinging the nodes, as the layout fetch is indirectly a measure of connectivity.
+            Map <String, ClusterStatusReport.ConnectivityStatus> connectivityStatusMap = getConnectivityStatusMap(layout);
+
+            log.debug("getClusterStatus: successful. Overall cluster status: {}", clusterStatus);
+
+            return new ClusterStatusReport(layout, clusterStatus, statusReliability, nodeStatusMap, connectivityStatusMap);
+        } catch (Exception e) {
+            log.error("getClusterStatus[{}]: cluster status unavailable. Exception: {}",
+                    this.runtime.getParameters().getClientId(), e);
+            return getUnavailableClusterStatusReport(layoutServers);
         }
-        return nodeStatusMap;
     }
 
-    /**
-     * Gets the cluster status by pinging each node.
-     * These pings are then compared to the layout to decide whether the cluster is:
-     * STABLE, DEGRADED OR UNAVAILABLE.
-     * The report consists of the following:
-     * Layout - at which the report was generated.
-     * ClientServerConnectivityStatusMap - the connectivity status of this client to the cluster.
-     * ClusterStatus - Health of the Corfu cluster.
-     *
-     * @return ClusterStatusReport
-     */
-    public ClusterStatusReport getClusterStatus() {
+    private ClusterStatusReport getUnavailableClusterStatusReport(List<String> layoutServers, ClusterStatusReliability reliability, @Nonnull Set<String> responsiveServers, Layout layout) {
+        return new ClusterStatusReport(layout,
+                ClusterStatus.UNAVAILABLE,
+                reliability,
+                layoutServers.stream().collect(Collectors.toMap(
+                        endpoint -> endpoint,
+                        node -> NodeStatus.NA)),
+                layoutServers.stream().collect(Collectors.toMap(
+                        endpoint -> endpoint,
+                        endpoint -> responsiveServers.contains(endpoint) ? ConnectivityStatus.RESPONSIVE : ConnectivityStatus.UNRESPONSIVE)));
+    }
+
+    private ClusterStatusReport getUnavailableClusterStatusReport(List<String> layoutServers) {
+        return getUnavailableClusterStatusReport(layoutServers, ClusterStatusReliability.UNAVAILABLE, Collections.emptySet(), null);
+    }
```

Comment:
can we pass something else but not null?

Reply:
what do you propose to pass? whenever this is the case, you could not connect to any layout server, hence there is no layout to take a decision on. If we pass a fake layout with all nodes unresponsive, I'm not sure how accurate this is, cause 1. it was not the layout on which this status is based on and 2. how could you know if this is not a client that came back from days down and is based on a very old layout, so we are showing out of date info...

Reply:
As discussed: please find out if we can use Optional instead of null